google-datacatalog-qlik-connector/src/google/datacatalog_connectors/qlik/scrape/base_engine_api_helper.py

```diff
@@ -0,0 +1,146 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import asyncio
+import json
+import logging
+import threading
+
+from urllib.parse import urlparse
+import websockets
+
+from google.datacatalog_connectors.qlik.scrape import constants
+
+
+class BaseEngineAPIHelper(abc.ABC):
+    """The base class for all Engine API Helpers."""
+
+    def __init__(self, server_address, auth_cookie):
+        # The server address starts with an http/https scheme. The below
+        # statement replaces the original scheme with 'wss', which is used for
+        # secure websockets communication.
+        self.__base_api_endpoint = f'wss://{urlparse(server_address).hostname}'
+        self.__auth_cookie = auth_cookie
+        self.__common_headers = {
+            constants.XRFKEY_HEADER_NAME: constants.XRFKEY,
+        }
+        self.__requests_counter = 0
+
+    def _connect_websocket(self, app_id):
+        """Opens a websocket connection.
+
+        Args:
+            app_id:
+              An App ID to be appended to the Engine API URLs when making
+              websocket connections and then get an isolated Qlik Engine
+              session for each Qlik App (see [Connecting to the Qlik Engine
+              JSON API](https://help.qlik.com/en-US/sense-developer/September2020/Subsystems/EngineAPI/Content/Sense_EngineAPI/GettingStarted/connecting-to-engine-api.htm)).  # noqa E510
+
+        Returns:
+            An awaiting function that yields a :class:`WebSocketClientProtocol`
+            which can then be used to send and receive messages.
+        """
+        uri = f'{self.__base_api_endpoint}/app/{app_id}' \
+              f'?Xrfkey={constants.XRFKEY}'
+
+        auth_cookie = self.__auth_cookie
+        headers = self.__common_headers.copy()
+        # Format the header value as <key>=<value> string.
+        headers['Cookie'] = f'{auth_cookie.name}={auth_cookie.value}'
+
+        return websockets.connect(uri=uri, extra_headers=headers)
+
+    def _generate_request_id(self):
+        self.__requests_counter += 1
+        return self.__requests_counter
+
+    @classmethod
+    def _run_until_complete(cls, future, max_seconds_to_wait):
+        loop = asyncio.get_event_loop()
+        timeout_controller = threading.Timer(
+            interval=max_seconds_to_wait,
+            function=cls._handle_event_loop_exec_timeout,
+            args=(loop,))
+        timeout_controller.start()
+        result = loop.run_until_complete(future)
+        timeout_controller.cancel()
+        return result
+
+    @classmethod
+    def _handle_event_loop_exec_timeout(cls, event_loop):
+        logging.warning(
+            'Timeout reached during the websocket communication session.')
+        event_loop.stop()
```

Comment:
I'm not an expert on `asyncio`, but checking some docs, for example, this stackoverflow post:
https://stackoverflow.com/questions/34710835/proper-way-to-shutdown-asyncio-tasks

If you call `event_loop.stop` it will stop waiting for the future, but the future (Task) will keep executing. 

Should you also cancel the task in someway? like something based on this stackoverflow post answer:
```
def shutdown():
    """Performs a clean shutdown"""
    log.info('received stop signal')
    for task in asyncio.Task.all_tasks():
        log.info(task)
        log.info('cancelling task')
        task.cancel()
```
let me know if this makes sense.

Reply:
It definitely does, thanks! I've limited the scope to cancel the tasks running in the current loop:
```
        for task in asyncio.Task.all_tasks(loop=event_loop):
            task.cancel()
```